### PR TITLE
Using transactions for all writes in the daemon scheduler

### PR DESCRIPTION
**Pull request by poojamaiya**
#### Summary
Changes to ensure all writes happen within a transaction

#### Implementation details
Using etcd's STM feature

#### Testing

- [X] daemon-scheduler binary built locally and unit-tests pass (`cd daemon-scheduler; make; cd ../`)
- [X] daemon-scheduler build in Docker succeeds (`cd daemon-scheduler; make release; cd ../`)

New tests cover the changes: yes

#### Description for the changelog
Using transactions for all writes in the daemon scheduler

#### Licensing

This contribution is under the terms of the Apache 2.0 License: Yes

#### Before merging

- [X] daemon-scheduler integration tests pass. Required setup details are listed [here](https://github.com/blox/blox/blob/dev/daemon-scheduler/internal/features/README.md).
- [X] daemon-scheduler end-to-end tests pass. Required setup details are listed [here](https://github.com/blox/blox/blob/dev/daemon-scheduler/internal/features/README.md).

### CODE REVIEW

File: daemon-scheduler/pkg/store/environment.go [hunk 1]
```diff
@@ -25,62 +26,74 @@ const (
...
+	key, err := generateEnvironmentKey(name)
+	if err != nil {
+		return nil, err
 	}
 
 	var environment types.Environment
-	environment.Name = name
 
-	key, err := generateEnvironmentKey(environment)
 	if err != nil {
```
Comment (kylbarnes) [inline daemon-scheduler/pkg/store/environment.go]: What err is this checking right here?

Comment (poojamaiya) [inline daemon-scheduler/pkg/store/environment.go]: Good catch! Thanks, missed deleting it.

File: daemon-scheduler/pkg/store/environment_test.go [hunk 2]
```diff
@@ -192,3 +213,24 @@ func (suite *EnvironmentTestSuite) TestList() {
 		assert.Contains(suite.T(), envs, expectedEnv, "Expected %s to be returned by ListEnvironments", expectedEnv)
 	}
 }
+
+func (suite *EnvironmentTestSuite) TestDeleteEnvironmentNoName() {
+	suite.datastore.EXPECT().Delete(gomock.Any(), gomock.Any()).Times(0)
+
+	err := suite.environmentStore.DeleteEnvironment(suite.ctx, "")
+	assert.NotNil(suite.T(), err, "Expected an error when deleting an environment with no name")
```
Comment (kylbarnes) [inline daemon-scheduler/pkg/store/environment_test.go]: Should these two be `assert.Error()` instead of `assert.NotNil()`?

Comment (poojamaiya) [inline daemon-scheduler/pkg/store/environment_test.go]: Sure, I think I've used NotNil() pretty much everywhere. I'll walk through the code and update.

File: daemon-scheduler/pkg/store/environment_stm_applier_test.go [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+	assert.Error(t, err, "Expected error when record key is not set in applier")
+}
+
+func TestValidateApplierNoValidateAndUpdateFunction(t *testing.T) {
+	applier := &EnvironmentSTMApplier{
+		key: "someKey",
+	}
+
+	err := applier.validateApplier()
+	assert.Error(t, err, "Expected error when record key is not set in applier")
```
Comment (kylbarnes) [inline daemon-scheduler/pkg/store/environment_stm_applier_test.go]: nit: message here should probably say "when validateAndUpdateFunction" is not set, or something along those lines.

File: daemon-scheduler/pkg/store/environment_stm_applier_test.go [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+		},
+		PutFunc: func(k string, val string, opts ...clientv3.OpOption) {
+			assert.Equal(t, key, k, "Unexpected key in Put")
+			assert.Equal(t, val, putEnvJSON, "Unexpected record in Put")
+		},
+	}
+
+	applier := &EnvironmentSTMApplier{
+		key: key,
+		validateAndUpdateEnv: func(exitingEnv *types.Environment) (*types.Environment, error) {
```
Comment (kylbarnes) [inline daemon-scheduler/pkg/store/environment_stm_applier_test.go]: nit, i think... should these four `exitingEnv` be `existingEnv`? I'm trying to figure out if this means that the existing environment is being passed in, or the exiting environment to be replaced. Normally i don't worry too much about typos, but I think it means two different things for `exiting` vs `existing`.

File: daemon-scheduler/pkg/store/environment_stm_applier_test.go (hunk 3 above)
Comment (poojamaiya) [inline daemon-scheduler/pkg/store/environment_stm_applier_test.go]: Fixed

File: daemon-scheduler/pkg/store/environment_stm_applier_test.go (hunk 4 above)
Comment (poojamaiya) [inline daemon-scheduler/pkg/store/environment_stm_applier_test.go]: It should be `existingEnv`. My bad! Fixed, thanks!

**Pull request merged**